### DISCUSSION TITLE
Persist issue/rehydrate

### DISCUSSION THREAD

**Discussion by asadbekabduvoitov-nordra** [pmndrs/zustand#2909] [Q&A] [answered]
Hello, guys. Just wanted to open a discussion for discussing a topic, I don't necessary want to say it as issue, because it's more of a use-case.  So I will briefly describe what I'm building and after that I will post codes as well. 

We're building an application, where user can have multiple organizations. In app we have **roles** page, where users can open roles by clicking into the card, and once they click it will be opened in new tab (inside application). The tabs functionality is almost same as browser tabs, where you can close, switch between tabs. And Im using zustand persist to store in the sessionStorage with dynamic key. I have seen/read a few discussions about dynamic key and how to rehydrate them. But the problem is switching between organization happens a lot faster, no refresh or anything needed. And opened tab needs to be stored on sessionStorage with organization id attached. 

```
const useTabs = create<RolesTab>()(
	persist(tabStore, {
		name: `${localStorage.getItem("active_org")}/tabs/roles`,
		storage: createJSONStorage(() => sessionStorage),
	}),
);
```

This is how I was creating stores, but the problem is unless I refresh the page, it was not working even when I rehydrate with setting new name. Name was being set, but even with new name it was returning previous organization's tab records. 

```
window.addEventListener("org-switch", () => {
       const storageKey = `${localStorage.getItem("active_org")}/tabs/roles`;
       useTabs?.persist?.setOptions({ name: storageKey });
       useTabs?.persist?.rehydrate();
});
```

This use case not worked for me. Not sure why, but unless I refresh the page, the tabs where mixed up with previous one. And I detected one odd behaviour, where it worked only when storage existed with the given name. Other wise it was using previous organization records. 

How I figured out:

```
const createTabStore = () => {
	return create<RolesTab>()(
		persist(tabStore, {
			name: `${localStorage.getItem("active_org")}/tabs/roles`,
			storage: createJSONStorage(() => sessionStorage),
		}),
	);
};

const rolesTabs: ReturnType<typeof createTabStore>[] = [];

export const useRolesTab = () => {
	const storageKey = `${localStorage.getItem("active_org")}/tabs/roles`;
	const foundStore = rolesTabs.find(
		({ persist }) => persist?.getOptions().name === storageKey,
	);

	if (!foundStore) rolesTabs.push(createTabStore());

	foundStore?.persist?.rehydrate();

	return foundStore?.()!;
};
```

This is how I figured out. By creating separate store for each organization. Just wanted to know that is this best practice or is there any way to optimize the way it's built.

**Comment by dbritto-dev** [comment, marked as answer]
@asadbekabduvoitov-nordra you could use `storage` events to keep everything in sync so you don't need to refresh/hydrate your store

Ref: https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event

**Reply by asadbekabduvoitov-nordra** [reply]
Yes, but the problem, is changes wont get reflected in the same tab. 

`addEventListener("storage", (event) => {});`

Wont get called in the same tab :( 

That's why I create custom event, and using custom event tried to rehydrate.

**Reply by asadbekabduvoitov-nordra** [reply]
But as for you comment, why do you think I don't need to rehydrate the store ?

Could you elaborate a bit or provide any examples ?

**Reply by dbritto-dev** [reply]
@asadbekabduvoitov-nordra like I said to keep everything in sync, the main tab still need to call `set` function. When I said `hydrate` I mean you don't need to refresh the tab in order to see the new changes. BTW, you could find some packages in the third-party libraries section -> https://github.com/pmndrs/zustand/blob/main/docs/integrations/third-party-libraries.md

**Reply by asadbekabduvoitov-nordra** [reply]
Makes sense, but how should I build `set` function ?

Or is this a function that zustand provides as a first setter function in the paramter of `create`.

**Reply by asadbekabduvoitov-nordra** [reply]
@dbritto-dev 
Thanks for the suggestion for taking a look to third-party libraries section, it's an art collection. I should definitely spend a bit more time to discover 🙌

**Reply by dbritto-dev** [reply]
@asadbekabduvoitov-nordra when I said `set` function is `set` (inside the state creator) and `setState` (from store itself)

**Reply by asadbekabduvoitov-nordra** [reply]
@dbritto-dev 

Yes, makes sense. But the problem is that our app uses clerk. And clerk doesn't provide any callback to be called when organization switch happened. Nor storage gets called in the current tab, so I can sync with sessionStorage. That's the tricky part, if that makes sense ?

**Reply by dbritto-dev** [reply]
@asadbekabduvoitov-nordra what about `afterSelectOrganizationUrl`? are you using `<OrganizationSwitcher />`?

**Reply by asadbekabduvoitov-nordra** [reply]
Hey, @dbritto-dev. Sorry for not being active over the past 5 days. Rebuild the store based on your suggestion, and it's working quite well, thanks. 

But, clerk doesn't seem to provide any callback, so what I did is:
- I have created custom even when organization changed. 
- I have faced with one issue, that even was being fired at any point, regardless of changing your active organization. 
- So I need to add one more check in order to find out, the even is fired for the actual organization change or for the session refresh or etc. 

It's working quite well. Thanks, @dbritto-dev.

**Reply by dbritto-dev** [reply]
Cool
